Update: I still haven't been able to use the update to build the newer Midori. I suspect the issue is the valaSupport but I'm not sure how to fix it:
```
[ 62%] Building C object tests/CMakeFiles/database.dir/database.c.o
[ 63%] Linking C executable database
[ 63%] Built target database
[ 63%] Generating download.c
/tmp/nix-build-midori-0.5.11.drv-0/midori-0.5.11/tests/download.vala:75.9-75.32: error: The name `listen_local' does not exist in the context of `Soup.Server'
        test_server.listen_local (0, Soup.ServerListenOptions.IPV4_ONLY);
        ^^^^^^^^^^^^^^^^^^^^^^^^
/tmp/nix-build-midori-0.5.11.drv-0/midori-0.5.11/tests/download.vala:80.16-80.35: error: The name `get_uris' does not exist in the context of `Soup.Server'
    var port = test_server.get_uris ().data.port;
               ^^^^^^^^^^^^^^^^^^^^
/tmp/nix-build-midori-0.5.11.drv-0/midori-0.5.11/tests/download.vala:80.9-80.48: error: var declaration not allowed with non-typed initializer
    var port = test_server.get_uris ().data.port;
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/tmp/nix-build-midori-0.5.11.drv-0/midori-0.5.11/tests/download.vala:88.53-88.56: error: The name `port' does not exist in the context of `download_properties'
    string test_url = "http://127.0.0.1:%u".printf (port);
                                                    ^^^^
Compilation failed: 4 error(s), 0 warning(s)
tests/CMakeFiles/download.dir/build.make:68: recipe for target 'tests/download_valac.stamp' failed
make[2]: *** [tests/download_valac.stamp] Error 1
CMakeFiles/Makefile2:2791: recipe for target 'tests/CMakeFiles/download.dir/all' failed
make[1]: *** [tests/CMakeFiles/download.dir/all] Error 2
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2
builder for ‘/nix/store/kp48n8rz6q67f4124dy7w4fa55ssnv3a-midori-0.5.11.drv’ failed with exit code 2
cannot build derivation ‘/nix/store/q6by0pi31vqkvcp9i99c9vmkyjly1j9n-midori-0.5.11.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/q6by0pi31vqkvcp9i99c9vmkyjly1j9n-midori-0.5.11.drv’ failed
```

Regardless, the issue should not affect non-Vala packages.  I suspect the solution is to place the vala bits under the dev output... But I'm not a vala developer so I can't write a small test unit to work on this. And seeing how Midori takes 3+ hours to compile...

AS such, the valaSupport should be considered a WIP feature. I could turn it off if asked. But it's harmless left as is so I won't unless asked to.

p.s. The pacakge can be tested or backported to stable by simply commenting out the outputs.

--- a/pkgs/development/libraries/libsoup/default.nix
+++ b/pkgs/development/libraries/libsoup/default.nix
@@ -1,33 +1,41 @@
 { stdenv, fetchurl, glib, libxml2, pkgconfig
-, gnomeSupport ? false, libgnome_keyring, sqlite, glib_networking, gobjectIntrospection
+, gnomeSupport ? true, libgnome_keyring, sqlite, glib_networking, gobjectIntrospection
+, valaSupport ? true, vala
 , libintlOrEmpty
 , intltool, python }:
 let
-  majorVersion = "2.52";
-  version = "${majorVersion}.2";
+  majorVersion = "2.54";
+  version = "${majorVersion}.1";
 in
 stdenv.mkDerivation {
   name = "libsoup-${version}";
 
   src = fetchurl {
     url = "mirror://gnome/sources/libsoup/${majorVersion}/libsoup-${version}.tar.xz";
-    sha256 = "1p4k40y2gikr6m8p3hm0vswdzj2pj133dckipd2jk5bxbj5n4mfv";
+    sha256 = "0cyn5pq4xl1gb8413h2p4d5wrn558dc054zhwmk4swrl40ijrd27";
   };
 
-  patchPhase = ''
+  prePatch = ''
     patchShebangs libsoup/
+  '' + stdenv.lib.optionalString valaSupport
+  ''
+     substituteInPlace libsoup/Makefile.in --replace "\$(DESTDIR)\$(vapidir)" "\$(DESTDIR)\$(girdir)/../vala/vapi"
   '';
 
   outputs = [ "dev" "out" ];
 
-  buildInputs = libintlOrEmpty ++ [ intltool python sqlite ];
+  buildInputs = libintlOrEmpty ++ [ intltool python sqlite ]
+    ++ stdenv.lib.optionals valaSupport [ vala ];
   nativeBuildInputs = [ pkgconfig ];
   propagatedBuildInputs = [ glib libxml2 gobjectIntrospection ]
     ++ stdenv.lib.optionals gnomeSupport [ libgnome_keyring ];
   passthru.propagatedUserEnvPackages = [ glib_networking.out ];
 
   # glib_networking is a runtime dependency, not a compile-time dependency
-  configureFlags = "--disable-tls-check --enable-vala=no" + stdenv.lib.optionalString (!gnomeSupport) " --without-gnome";
+  configureFlags = "--disable-tls-check"
+    + stdenv.lib.optionalString (!valaSupport) " --enable-vala=no"
+    + stdenv.lib.optionalString (valaSupport) " --enable-vala=yes"
+    + stdenv.lib.optionalString (!gnomeSupport) " --without-gnome";
 
   NIX_CFLAGS_COMPILE = stdenv.lib.optionalString stdenv.isDarwin "-lintl";
 

